### ISSUE TITLE
Support library dependencies

### ISSUE THREAD

**Issue by jhpedemonte** [maqetta/maqetta#1105]
(Thought there was a bug open already, but can't find it.)

Need to add support for a library depending on another.

---

Proposal: https://github.com/maqetta/maqetta/wiki/Library-Packages

**Comment by jhpedemonte** [comment]
We should start using the CommonJS Package spec ([v1.0](http://wiki.commonjs.org/wiki/Packages/1.0) & [v1.1](http://wiki.commonjs.org/wiki/Packages/1.1)) as the top level metadata for libraries.  Kris Zyp maintains a [template](https://github.com/kriszyp/commonjs-package-template/blob/master/package.json) that we can start with.

This spec is starting to be used by others and provides a way to specify dependencies between libraries.  It also defines some of the same metadata that we have at the top of our **widgets.json** files.

Note that to the CommonJS Package spec, Kris also adds the [Mappings proposal](http://wiki.commonjs.org/wiki/Packages/Mappings/C).

In order to support our metadata, we can add support for an extra member to the `dependencies` top-level property of **package.json**.  Something like `metadata` or `widgets`, although I'm not completely sure about either of those names.

This extra property would point to the directory which contains the **widgets.json** file, which now would only consist of an ordered array of widget descriptors.

---
- This essentially splits the data that we currently have in widgets.json into two separate files.  Unfortunately, this increases the number of network requests from the client.  How can we improve on this situation?
- I think we should split the Dojo library into it's separate components (Dojo, Dijit, Dojox Mobile).  Note that this follows what is happening in the [Dojo Foundation repository](http://packages.dojofoundation.org/list.html), where Dojo and Dijit are separate packages.
  
  This would solve the issue outlined in the [foundation data structures](https://github.com/maqetta/maqetta/wiki/Foundation-data-structures) wiki, of "A particular library can contain one or more widget collections".  Instead of supporting multiple widget collections in one library, each widget collection would be its own package.  So we would have separate Dijit and Dojox Mobile packages, each which depends on the (non-widget) Dojo package. (Does Dojox Mobile depend on Dijit?)

@JonFerraiolo @peller

**Comment by jhpedemonte** [comment]
The big advantage of simply adding a `metadata`/`widgets` prop under `directories` is that we can take an existing library which has a **package.json** and add that line and the metadata directory and it's set for use in Maqetta.

**Comment by jhpedemonte** [comment]
An example, of the Dojox Mobile package:

``` javascript
{
  "name": "dojox-mobile",
  "description": "Dojox Mobile widgets",
  "version": "1.7",
  "licenses": [
     {
         "type": "AFLv2.1",
         "url": "http://trac.dojotoolkit.org/browser/dojo/trunk/LICENSE#L43"
     },
     {
         "type": "BSD",
         "url": "http://trac.dojotoolkit.org/browser/dojo/trunk/LICENSE#L13"
     }
  ],
  "bugs": "http://bugs.dojotoolkit.org/",
  "keywords": ["Dojo", "Mobile"],
  "homepage": "http://dojotoolkit.org/",
  "dependencies": {
    "dojo": "1.7"
  },
  "mappings": {
    "dojo": "https://github.com/dojo/dojo/zipball/v1.7.0"
  },
  "directories": {
    "lib": ".",
    "metadata": "metadata"
  },
  "icon": "http://dojofoundation.org/images/icons/5.png"
}
```

**Comment by jhpedemonte** [comment]
An idea on how to decrease performance hit from multiple network requests: when the client requests the library metadata (package.json), the server can merge multiple library files and send that back instead.  So client would get back one file consisting of the library's package.json, widget.json and library.js (the JS code for the library, formerly "callbacks.js").  We could do similar things when requesting a widget's metadata -- send back one file consisting of widget OAM and any helper JS files.

The server would generate the merged files once and cache the result per library.

@childsb - Does that sound good to you?

**Comment by JonFerraiolo** [comment]
Users hardly even will be adding new libraries. It's a .001% event, where the other 99.999% of the time users will either use the libraries that are there by default or use the libraries that were there by default plus a library they added themselves. Therefore, we don't want to punish the world from a performance perspective by doing N different network requests for each widget package with each user session. We should optimize performance for teh 99.999% case.

One possibility is that server consolidates all of the widgets.json and OAM files into a single JSON file delivered as a single web request to the client whenever: (1) new project is created, or (2) user modifies the list of libraries.

**Comment by peller** [comment]
taking things a bit further, we shouldn't have to distribute code with our plugins anymore, just a package.json or, I suppose if everything we needed was there, reference to a package.json

**Comment by JonFerraiolo** [comment]
Going even further, can't we get rid of our widget library plugins if our server code looks at a particular set of locations for widget package files (ie.., package.json) within a particular site folder, (organization folder depending on if we support organizations), group folder and user folder? The widget package files then could either point to the appropriate JavaScript file within the same server or on a remote server, right? If on a remote server, probably point to a ZIP file.

**Comment by peller** [comment]
yup.  Mappings should point at a zip file.  The spec says it should have an "archive" property.  Kris' example uses the "dojo" property?

**Comment by jhpedemonte** [comment]
Right, so we should get rid of the JAR bundles we have now and just have a directory of libraries, which have a **package.json** at the top level.  If we want to allow "external" (3rd party libraries), then we can allow the user or the server admin to add a URL that points either to (1) an archive which when expanded contains a directory with package.json at the top-level; or (2) a server location where adding "/package.json" to the URL is a valid URL.

@peller, the `mappings` property is an addition to go along with the `dependencies` prop.  When the spec was first created (and became a "draft" in v1.0), there wasn't anything that specified where the dependencies are located.  So in the example above, there is a "dojo" dependency, and `mappings` specifies where it is located.

**Comment by peller** [comment]
there is a Mappings/C spec at CommonJS.  Thought that was Kris'.  I'm sure it's morphed and continues to do so.  The example implies that mappings itself is a named map for libraries, perhaps allowing 1-library-to-many archives. (actually, C is the exception... B and D seem to be the ones that work this way)

**Comment by peller** [comment]
even the package json itself could be remote, if packaged in a JSONP or AMD manner

**Comment by jhpedemonte** [comment]
The **Mappings/C** proposal seems to be the one in use by both Kris and the PINF loader.

**Comment by jhpedemonte** [comment]
What do you guys think about splitting up our current Dojo bundle into 3 separate bundles: (non-widget) **Dojo**, **Dijit** and **Dojox Mobile**?  With the latter 2 having a dependency on the former?

**Comment by peller** [comment]
Seems like a good idea, for the reasons you mention.  Does Dojo actually have any 'widget' entries for the palette?

**Comment by jhpedemonte** [comment]
No, just Dijit and Dojox Mobile.

**Comment by peller** [comment]
Even a better reason to split them out.

**Comment by jhpedemonte** [comment]
Actually, the division wouldn't be as clean as I assumed -- we have quite a few non-mobile Dojox widgets.  We support DataGrid and TimeSpinner, but the rest are in the "untested" category.

So really, the divisions would be more like: Dojo, Dijit, Dojox, Dojox Mobile.

**Comment by peller** [comment]
dojox will break apart eventually, most often along "subproject" lines, like dojox.mobile.  Some subprojects, like dojox.form or dojox.widget are messy and will probably result in separate packages, and various other widgets are likely to be abandoned.  Splitting out dijit and dojox.mobile for starters seems reasonable.

**Comment by jhpedemonte** [comment]
Proposal for handling library and metadata packages: https://github.com/maqetta/maqetta/wiki/Library-Packages

**Comment by peller** [comment]
Looks good, though I was thinking about maybe adding the metadata files as properties on the package.json file so there would be an explicit reference to find them, rather than using some naming convention for directories.  Perhaps it's as simple as adding two fields in addition to everything you have:

   widget-metadata: './library/metadata/widget.json'

**Comment by JonFerraiolo** [comment]
(1) I'm ok with either Adam's proposal for adding fields to point to maqetta metadata files or using a file naming convention. However, I would think it's easier and simpler for developers if we tell them to adopt a naming convention.  Otherwise, it's just yet another field to figure out and fill out. We are already following the naming convention pattern with other things such as _.theme files, and the OAM spec actually talks about the ```__oam.xml``` naming convention. Whatever, we should follow the same approach (ie, field in index file pointing to filename vs naming convention) across all of our different file types.

Other comments/questions:
- (2) Would it be better to have separate folders for the OAM metadata and the Maqetta metadata? Perhaps `oameta/` and `maqmeta/`?
- (3) Will developers ever need to have multiple maqetta metadata files (i.e., widgets.json)? Right now our dijit widgets.json file is pretty unwieldy. Most of the time the server will consolidate, so why not allow multiple files and have the server search just like we do for themes.
- (4) Is there an error in case 3 where the first dependency points to "library-maq". Shouldn't this dependency point to "library-oam"?
- (5) In the section on Conditional Dependencies, first paragraph, it says `order determines priority`. Sorry about my JS ignorance, but with associative arrays, does JS preserve property order?
- (6) Does the package spec encourage developers to somehow namespace-quality the 'name' values for their package? I'm just wondering if there are potential future name collisions if things aren't namespace-qualified.

**Comment by peller** [comment]
Good point.  JS does not preserve property order.  In reality, most implementations do, but the spec does not.

**Comment by JonFerraiolo** [comment]
so we need to use an array instead of an object

**Comment by jhpedemonte** [comment]
Note that the file is **JSON**, not **JS**.  So order is significant.  (JSON != JS)

**Comment by jhpedemonte** [comment]
The Packages spec doesn't mention any naming conventions.  However, the "metadata" property is used for both the OAM metadata and the Maqetta metadata ("widgets.json").  For the OAM metadata, there is no single file, but a collection, mimicking the directory structure of the library files.  For these reasons, we should leave it pointing to a directory in the package.

Issue (2): I don't think it really matters.  The server should be able to map to the correct location, whether the OAM and Maqetta metadata files are in the same or different locations.

Issue (3): I'd rather keep it simple and just allow one file.  I'm open to changing my mind based on user feedback.

Issue (4): Fixed.

Issue (5): Answered above.

Issue (6) "namespaces": The spec makes no mention of namespacing property names.  It does mention names reserved for future use.  The existing Dojo libraries do add a property extension ("dojoBuild"), which is namespaced.  Just to avoid possible collisions, it's best to namespace.

However, in my proposal I only added the "metadata" property (change to "widgets" or "widgets-metadta"), which is meant to point to both the OAM and Maqetta metadata.  Since this is pushing the OAM format for widget metadata, I don't think we should namespace -- encourage others to make use of it.

**Comment by peller** [comment]
Confused on the ordering issue.  If you're going to use read/write the JSON file using JS data structures and JSON.stringify/parse, you're not going to be able to preserve order.  You could go out of your way to handle the order of these strings in a JSON file, but it would require special logic, no?  Seems like that would be very awkward.

**Comment by jhpedemonte** [comment]
No, you are right -- JSON is converted into JS, where ordering of object properties is unspecified.  We should bring it up, maybe contact Kris about it.

**Comment by jhpedemonte** [comment]
OK, I rewrote the [proposal](https://github.com/maqetta/maqetta/wiki/Library-Packages) 5-10 times since Friday, but I think I finally figured it out.  I added references that I used to create the proposal.  Incidently, yes, the `dependencies` property does support choices between 2 or more deps.

Also added theme metadata support.  I went with specifying a directory, under which could be found `*.theme` files.  Is that correct?

**Comment by JonFerraiolo** [comment]
Right now we have a themes/ folder, inside of which we have a folder for each available theme (e.g., themes/claro, themes/sketch, themes/android).

I don't think this affects the packaging question, but I just wanted to remind everyone that I'd like to address #144 so that there is a folder holding the untouched original theme content (e.g., claro/). This is for 2 reasons: (1) it's always better to leave packaged content untouched (versus stuffing extra files inside of that package), (2) the Eclipse guys want to allow the user to double-click on *.theme files to launch Maqetta theme editor and it's harder to fine the *.theme files when inside a folder with a zillion CSS files in the same folder.

**Comment by JonFerraiolo** [comment]
oh, yeah, I forgot, proposal looks good!

**Comment by peller** [comment]
@jhpedemonte: so it sounds like dependencies needs to be an array.  Which one of us should post the question to commonjs?  Shall we draw straws?

@JonFerraiolo: I strongly agree with both points regarding #144.  Then we wouldn't even have to do a search for *.theme files anymore, except for that one directory.

**Comment by peller** [comment]
Looks good.  @jhpedemonte: you left out a permutation or two, such as widget + maqetta metadata being packaged together (conceivable)  No problem, right?  Just a different mixin combination?

**Comment by jhpedemonte** [comment]
As mentioned on the wiki, it was long established in CommonJS that almost all (or actually all?) JS implementations keep the order of properties as the are specified on an object, and that ES5 will codify this in the spec.  Therefore, it was deemed that an object, rather than an array, would work just fine and was preferable (more terse than array).  So we're stuck with what we've got.

Yes, there are several other permutations not listed (less likely to come up than the ones listed).  As long as we handle the overlays and dependencies correctly, it shouldn't be an issue.

**Comment by jhpedemonte** [comment]
Still left to do in M5:
- If library A depends on library B, make sure the user cannot disable library B without disabling library A (#1386).
- (Maybe) If a library A depends on library B, automatically add B's main JS file when first adding a widget from library A.  Not sure how much we need this -- for example, for Drawing Tools & Clipart libraries (which depend on Dojo), the individual OAM files contain a statement requiring dojo.js.  That may be enough.

**Comment by jhpedemonte** [comment]
-> M7

**Comment by JonFerraiolo** [comment]
@aerwin 

Was this issue related to gridx's needs in order to get gridx working in a build?

**Comment by aerwin** [comment]
No, gridx's needs are covered in #2067.

**Comment by jhpedemonte** [comment]
-> M8

**Comment by jhpedemonte** [comment]
-> M9
